lib/apiservers/service/restapi/handlers/common.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/cmd/vic-machine/common"
+	"github.com/vmware/vic/lib/apiservers/service/restapi/handlers/util"
+	"github.com/vmware/vic/lib/config"
+	"github.com/vmware/vic/lib/install/data"
+	"github.com/vmware/vic/lib/install/management"
+	"github.com/vmware/vic/lib/install/validate"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/version"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+type buildDataParams struct {
+	target          string
+	thumbprint      *string
+	datacenter      *string
+	computeResource *string
+}
+
+func buildData(ctx context.Context, params buildDataParams, principal interface{}) (*data.Data, error) {
+	d := data.Data{
+		Target: &common.Target{
+			URL: &url.URL{Host: params.target},
+		},
+	}
+
+	if c, ok := principal.(Credentials); ok {
+		d.Target.User = c.user
+		d.Target.Password = &c.pass
+	} else {
+		d.Target.CloneTicket = principal.(Session).ticket
+	}
+
+	if params.thumbprint != nil {
+		d.Thumbprint = *params.thumbprint
+	}
+
+	if params.datacenter != nil {
+		validator, err := validateTarget(ctx, &d)
```

Comment:
Yes and no; we're using `validateTarget` to validate that the target is valid without considering the datacenter _and_ using it to resolve the datacenter parameter.

I agree this could be cleaner. It's a bit of a stretch given the current title, but I think this would be best addressed as a part of the work in #6032.